src/Solver/Maxwell.h

```diff
@@ -0,0 +1,85 @@
+//
+// Class Maxwell
+//   Base class for solvers for Maxwell's equations
+//
+
+#ifndef IPPL_MAXWELL_H
+#define IPPL_MAXWELL_H
+
+#include "Types/Vector.h"
+
+#include "Field/Field.h"
+
+#include "FieldLayout/FieldLayout.h"
+#include "Meshes/UniformCartesian.h"
+
+namespace ippl {
+
+    template <typename EMField, typename SourceField>
+    class Maxwell {
+    public:
+        // typedefs for the different fields and vector fields
+        using typeR = typename SourceField::value_type;
+
+        constexpr static unsigned Dim = EMField::dim;
+        typedef typename EMField::Mesh_t Mesh_t;
+        typedef typename EMField::Centering_t Centering_t;
+        typedef Vector<typeR, Dim> Vector_t;
+        typedef Field<Vector_t, Dim, Mesh_t, Centering_t> VectorSourceField_t;
+
+        // define type for field layout
+        typedef FieldLayout<Dim> FieldLayout_t;
+
+        // type for communication buffers
+        using memory_space = typename SourceField::memory_space;
+        using buffer_type  = Communicate::buffer_type<memory_space>;
+
+        /*!
+         * Default constructor for Maxwell solvers;
+         */
+        Maxwell() {}
+
+        Maxwell(SourceField& charge, VectorSourceField_t& current, EMField& E, EMField& B) {
+            setSources(charge, current);
+            setEMFields(E, B);
+        }
+
+        /*!
+         * Set the problem RHS (charge & current densities)
+         * @param charge Reference to rho field
+         * @param current Reference to J field
+         */
+        virtual void setSources(SourceField& charge, VectorSourceField_t& current) {
+            rhoN_mp = &charge;
+            JN_mp   = &current;
+        }
+
+        /*!
+         * Set the problem LHS (electromagnetic fields)
+         * @param E Reference to electric field
+         * @param B Reference to magnetic field
+         */
+        virtual void setEMFields(EMField& E, EMField& B) {
+            En_mp = &E;
+            Bn_mp = &B;
+        }
+
+        /*!
+         * Solve the electromagnetic problem (Maxwell's eqs)
+         */
+        virtual void solve() = 0;
+
+        virtual ~Maxwell() {}
+
+    protected:
+        // fields containing reference to charge and current
+        SourceField* rhoN_mp       = nullptr;
```

Comment:
why not assigning the nullprt in the constructor?

Reply:
Consider smart pointers.

Reply:
In Poisson.h, we use normal pointers for rhs_mp and lhs_mp, and then for example in the FFTPoissonSolver the mesh and layout of the rhs_mp is also a normal pointer, but all other pointers are smart. I do not know why this is like this, but if I change to smart pointers here, it would make sense to change everywhere in the solvers. Should I do that?

Reply:
Actually after reading a bit more on it, I think we should leave all the solver class members which will be pointers to fields coming from the outside (e.g. rhs and lhs) as normal pointers, not smart, since otherwise we need to allocate using `new` to initialize rhs and lhs. What do you think?